applications/sdp/mspi/src/main.c

```diff
@@ -52,215 +57,494 @@ typedef struct __packed {
+	nrf_vpr_csr_vio_dir_set(dir);
+	nrf_vpr_csr_vio_out_set(out);
+
+	nrf_vpr_csr_vio_mode_out_set(&out_mode);
+	nrf_vpr_csr_vio_mode_in_buffered_set(NRF_VPR_CSR_VIO_MODE_IN_CONTINUOUS);
+
+	nrf_vpr_csr_vio_config_get(&config);
+	config.input_sel = false;
+	nrf_vpr_csr_vio_config_set(&config);
+
+	/* Counter settings */
+	nrf_vpr_csr_vtim_count_mode_set(0, NRF_VPR_CSR_VTIM_COUNT_RELOAD);
+	/* TODO: Jira ticket: NRFX-6703
+	 *       Top value of VTIM. This will determine clock frequency
+	 *                         (SPI_CLOCK ~= CPU_CLOCK / (2 * TOP)).
+	 *       Calculate this value based on frequency
+	 */
+	nrf_vpr_csr_vtim_simple_counter_top_set(0, 4);
+
+	/* Set number of shifts before OUTB needs to be updated.
+	 * First shift needs to be increased by 1.
+	 */
+	nrf_vpr_csr_vio_shift_cnt_out_set(BITS_IN_WORD / out_mode.frame_width);
+	nrf_vpr_csr_vio_shift_cnt_out_buffered_set(BITS_IN_WORD / out_mode.frame_width - 1);
+
+	uint8_t last_word_length = data_length % BITS_IN_WORD;
+	uint8_t penultimate_word_length = BITS_IN_WORD;
+
+	xfer_ll_params.words = NRFX_CEIL_DIV(data_length, BITS_IN_WORD);
+	xfer_ll_params.last_word = ((uint32_t *)xfer_ll_params.data)[xfer_ll_params.words - 1];
+
+	/* Due to hardware limitations it is not possible to send only 1 clock cycle.
+	 * Therefore when data_length%32==1  last word is sent shorter (24bits)
+	 * and the remaining byte and 1 bit is sent together.
+	 */
+	if (last_word_length == 0) {
+
+		last_word_length = BITS_IN_WORD;
+		xfer_ll_params.last_word =
+			((uint32_t *)xfer_ll_params.data)[xfer_ll_params.words - 1];
+
+	} else if ((last_word_length / out_mode.frame_width == 1) && (xfer_ll_params.words > 1)) {
+
+		penultimate_word_length -= BITS_IN_BYTE;
+		last_word_length += BITS_IN_BYTE;
+		xfer_ll_params.last_word =
+			((uint32_t *)xfer_ll_params.data)[xfer_ll_params.words - 2] >>
+				(BITS_IN_WORD - BITS_IN_BYTE) |
+			((uint32_t *)xfer_ll_params.data)[xfer_ll_params.words - 1] << BITS_IN_BYTE;
+	}
+
+	xfer_ll_params.last_word_clocks = last_word_length / out_mode.frame_width;
+	xfer_ll_params.penultimate_word_clocks = penultimate_word_length / out_mode.frame_width;
+
+	if (xfer_ll_params.words == 1) {
+		nrf_vpr_csr_vio_shift_cnt_out_set(xfer_ll_params.last_word_clocks);
+	} else if (xfer_ll_params.words == 2) {
+		nrf_vpr_csr_vio_shift_cnt_out_set(xfer_ll_params.penultimate_word_clocks);
+	}
 }
 
-void configure_clock(enum mspi_cpp_mode cpp_mode)
+static void configure_clock(enum mspi_cpp_mode cpp_mode)
 {
 	nrf_vpr_csr_vio_config_t vio_config = {
 		.input_sel = 0,
 		.stop_cnt = 0,
 	};
-
-	nrf_vpr_csr_vio_dir_set(PIN_DIR_OUT_MASK(VIO(NRFE_MSPI_SCK_PIN_NUMBER)));
+	uint16_t out = nrf_vpr_csr_vio_out_get();
 
 	switch (cpp_mode) {
 	case MSPI_CPP_MODE_0: {
 		vio_config.clk_polarity = 0;
-		nrf_vpr_csr_vio_out_set(PIN_OUT_LOW_MASK(VIO(NRFE_MSPI_SCK_PIN_NUMBER)));
+		out = BIT_SET_VALUE(out, pin_to_vio_map[NRFE_MSPI_SCK_PIN_NUMBER],
```

Comment:
This CLK pin mapping should be placed in NRFE_MSPI_CONFIG_PINS also.
Should be mapped by pin function not hard coded.